src/libstd/collections/hash/table.rs

```diff
@@ -129,16 +119,7 @@ struct GapThenFull<K, V, M> {
 
 /// A hash that is not zero, since we use a hash of zero to represent empty
 /// buckets.
-#[derive(PartialEq, Copy)]
-pub struct SafeHash {
-    hash: u64,
-}
-
-impl SafeHash {
-    /// Peek at the hash value, which is guaranteed to be non-zero.
-    #[inline(always)]
-    pub fn inspect(&self) -> u64 { self.hash }
-}
+pub type SafeHash = NonZero<u64>;
```

Comment:
I don't see anything particularly wrong with just having `0` or `null` be a sentinel value. Not everything needs to be encoded in types. Seems to just make more code.